File: src/牛客网/剑指offer/java/MaxInWindows.java
package 牛客网.剑指offer.java;

import java.util.ArrayList;

/**
 * @since 2020/7/15 11:00
 * <p>
 * 给定一个数组和滑动窗口的大小，找出所有滑动窗口里数值的最大值。
 */
public class MaxInWindows {

    public ArrayList<Integer> maxInWindows(int[] num, int size) {
        ArrayList<Integer> ret = new ArrayList<>();
        if (size == 0 || size < 1 || num.length < size) return ret;
        for (int i = 0; i < num.length - (size - 1); ++i) { // 此处注意i的最大值要小于 num.length - (size - 1)，最后一个窗口。
            int max = Integer.MIN_VALUE;
            for (int j = 0; j < size; j++) {
                max = Math.max(max, num[i + j]);
            }
            ret.add(max);
        }
        return ret;
    }
}

File: src/offer_kotlin/MaxInWindows.kt
package offer_kotlin

import java.util.*

/**
 * @since 2020/7/15 11:16
 */
class MaxInWindows {

    fun maxInWindows(num: IntArray, size: Int): ArrayList<Int>? {
        val ret = ArrayList<Int>()
        if (size == 0 || size < 1 || num.size < size) return ret
        for (i in 0 until num.size - (size - 1)) { // 此处注意i的最大值要小于 num.length - (size - 1)，最后一个窗口。
            var max = Int.MIN_VALUE
            for (j in 0 until size) {
                max = max.coerceAtLeast(num[i + j])
            }
            ret.add(max)
        }
        return ret
    }
}